File: frontend-design/scripts/contrast_check.py
#!/usr/bin/env python3
"""Check WCAG contrast ratio for two hex colors.

Usage:
  python scripts/contrast_check.py '#111827' '#ffffff'
  python scripts/contrast_check.py '#777' '#fff' --large
"""

from __future__ import annotations

import argparse
import re


def parse_hex(value: str) -> tuple[int, int, int]:
    raw = value.strip().lstrip("#")
    if not re.fullmatch(r"[0-9a-fA-F]{3}|[0-9a-fA-F]{6}", raw):
        raise ValueError(f"Invalid hex color: {value}")
    if len(raw) == 3:
        raw = "".join(ch * 2 for ch in raw)
    return int(raw[0:2], 16), int(raw[2:4], 16), int(raw[4:6], 16)


def channel(v: int) -> float:
    x = v / 255
    return x / 12.92 if x <= 0.03928 else ((x + 0.055) / 1.055) ** 2.4


def luminance(rgb: tuple[int, int, int]) -> float:
    r, g, b = [channel(v) for v in rgb]
    return 0.2126 * r + 0.7152 * g + 0.0722 * b


def ratio(fg: str, bg: str) -> float:
    l1 = luminance(parse_hex(fg))
    l2 = luminance(parse_hex(bg))
    lighter = max(l1, l2)
    darker = min(l1, l2)
    return (lighter + 0.05) / (darker + 0.05)


def main() -> int:
    parser = argparse.ArgumentParser(description="Check WCAG contrast ratio")
    parser.add_argument("foreground")
    parser.add_argument("background")
    parser.add_argument("--large", action="store_true", help="Evaluate as large text")
    args = parser.parse_args()
    value = ratio(args.foreground, args.background)
    aa = value >= (3.0 if args.large else 4.5)
    aaa = value >= (4.5 if args.large else 7.0)
    ui = value >= 3.0
    print(f"Contrast: {value:.2f}:1")
    print(f"UI components AA (3:1): {'PASS' if ui else 'FAIL'}")
    print(f"Text AA ({'3' if args.large else '4.5'}:1): {'PASS' if aa else 'FAIL'}")
    print(f"Text AAA ({'4.5' if args.large else '7'}:1): {'PASS' if aaa else 'FAIL'}")
    return 0 if aa else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend-design/scripts/contrast_check.mjs
#!/usr/bin/env node
/** Check WCAG contrast ratio for two hex colors.
 *
 * Usage:
 *   node scripts/contrast_check.mjs '#111827' '#ffffff'
 *   node scripts/contrast_check.mjs '#777' '#fff' --large
 */

import { pathToFileURL } from "node:url";

export function parseHex(value) {
  let raw = String(value).trim().replace(/^#/, "");
  if (!/^(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/.test(raw)) throw new Error(`Invalid hex color: ${value}`);
  if (raw.length === 3) raw = [...raw].map((character) => character.repeat(2)).join("");
  return [0, 2, 4].map((offset) => Number.parseInt(raw.slice(offset, offset + 2), 16));
}

function channel(value) {
  const normalized = value / 255;
  return normalized <= 0.03928 ? normalized / 12.92 : ((normalized + 0.055) / 1.055) ** 2.4;
}

export function luminance(rgb) {
  const [red, green, blue] = rgb.map(channel);
  return 0.2126 * red + 0.7152 * green + 0.0722 * blue;
}

export function ratio(foreground, background) {
  const first = luminance(parseHex(foreground));
  const second = luminance(parseHex(background));
  return (Math.max(first, second) + 0.05) / (Math.min(first, second) + 0.05);
}

export function main(args = process.argv.slice(2)) {
  const largeIndex = args.indexOf("--large");
  const large = largeIndex !== -1;
  const colors = args.filter((argument) => argument !== "--large");
  if (colors.length !== 2) {
    console.error("Usage: contrast_check.mjs <foreground> <background> [--large]");
    return 2;
  }
  try {
    const value = ratio(colors[0], colors[1]);
    const aaThreshold = large ? 3 : 4.5;
    const aaaThreshold = large ? 4.5 : 7;
    console.log(`Contrast: ${value.toFixed(2)}:1`);
    console.log(`UI components AA (3:1): ${value >= 3 ? "PASS" : "FAIL"}`);
    console.log(`Text AA (${aaThreshold}:1): ${value >= aaThreshold ? "PASS" : "FAIL"}`);
    console.log(`Text AAA (${aaaThreshold}:1): ${value >= aaaThreshold ? "PASS" : "FAIL"}`);
    return value >= aaThreshold ? 0 : 1;
  } catch (error) {
    console.error(error.message);
    return 2;
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) process.exitCode = main();
